### DISCUSSION TITLE
RTR admin execute_admin_command with put command failing

### DISCUSSION THREAD

**Discussion by usedtire** [CrowdStrike/falconpy#431] [Q&A] [answered]
When using the falconpy module to run the put command of an exe available within my put files it fails.  I can run the command "put text.exe" directly in RTR session and it will download it.  Then every other command I have tried to execute with falconpy works.  

falcon_rtra.execute_admin_command(base_command="put", command_string="put test.exe", session_id=session_id, persistent=True)

Any insight into what the problem might be?

**Comment by jshcodes** [comment]
Hi @usedtire -

So when you upload directly from your machine it works, but when you try from your cloud files it fails?

If you adjust your code to the following, does it upload properly?

```python
falcon_rtra.execute_admin_command(base_command="put",
                                  command_string="put -CloudFile='test.exe'",
                                  session_id=session_id,
                                  persistent=True
                                  )
```

(Or did I misunderstand, and you're trying to execute it from your cloud put files?)

**Reply by usedtire** [reply]
When I use the web UI and launch the RTR Session to the system I can run the command "put test.exe" and it pulls from my available put files.  When I attempt the above command nothing happens.  Even modifying the command string as you have shown has not worked.

**Reply by jshcodes** [reply]
Copy that! 

Will attempt to recreate this issue and will report back once we've found something. 🔬

**Comment by jshcodes** [comment, marked as answer]
So this works in my test environment, but I've not yet tested on a Windows target and it's just a JPG file.  Could you give this a shot and let us know the result?

> Note: The put operation is dropping the file on the root of the file system using the following code.

```python
import os
from falconpy import RealTimeResponse, RealTimeResponseAdmin, OAuth2, Hosts

auth = OAuth2(client_id=os.environ["FALCON_CLIENT_ID"],
              client_secret=os.environ["FALCON_CLIENT_SECRET"]
              )
hosts = Hosts(auth_object=auth)
rtr = RealTimeResponse(auth_object=auth)
rtr_admin = RealTimeResponseAdmin(auth_object=auth)

target_host="TARGET_HOSTNAME_HERE"  # You can append a wildcard here if necessary

aid = hosts.query_devices_by_filter(filter=f"hostname:'{target_host}'")["body"]["resources"][0]

session_id = rtr.init_session(device_id=aid)["body"]["resources"][0]["session_id"]
print(f"Session ID: {session_id}")

filename = "testfile.jpg"
testfile = [('file', (filename, open(filename, 'rb').read(), 'image/jpg'))]

upload_result = rtr_admin.create_put_files(name=filename,
                                           description="Testing file",
                                           files=testfile
                                           )
if upload_result["status_code"] == 409:
    print("File already exists in Cloud repository")
elif upload_result["status_code"] == 200:
    print("File successfully uploaded")
else:
    print("An error occurred.")


put_file_id = rtr_admin.list_put_files(filter=f"name:'{filename}'")["body"]["resources"][0]
print(f"Put file ID: {put_file_id}")

# Put the file on the target machine
rtr_admin.execute_admin_command(base_command="put",
                                command_string=f"put {filename}",
                                session_id=session_id,
                                persist=True
                                )

# Cleanup our session
rtr.delete_session(session_id=session_id)
```

After running this once, you should be able to comment out lines 17 - 28 and then delete the local file (or rename it). Running the script after doing so should still put the file on the target system.

**Reply by jshcodes** [reply]
Also appears to be working on a Windows target (and dropped the file on the root of the file system.)

**Reply by usedtire** [reply]
I just copied your code over, inserted my ID and Secret as well as the hostname.  This is the error I received.  I apologize but I'm fairly new to using OAuth2, but any idea why I would be getting this?

Traceback (most recent call last):
  File "testsample.py", line 4, in <module>
    auth = OAuth2(client_id=os.environ["(KEY DELETED)"],
  File "E:\Python3\lib\os.py", line 678, in __getitem__
    raise KeyError(key) from None
KeyError: '(KEY DELETED)'

**Reply by jshcodes** [reply]
This looks like it's crashing looking for the environment variables I'm using to pass Client ID / Secret.

You can update the code to hardcode these values for your testing (completely remove "os.environ[...]") or create environment variables for **FALCON_CLIENT_ID** and **FALCON_CLIENT_SECRET**. 

> You might have to reload your terminal to get the new environment variables depending on how you create them.

**Reply by usedtire** [reply]
Everything worked except the put to the host.  It authed me, created the file on CS and created a session, but did not put the file to the host I provided. Could this be a permissions issue with the API?  Is there a specific permission I need on my key for a put?

**Reply by jshcodes** [reply]
Potentially.  Do you have READ & WRITE to RTR and RTR Admin?

Also, you've confirmed the file isn't showing up in C:\ ? (File system root.)

**Reply by jshcodes** [reply]
This version is expanded a bit so we can check the status of the put command.  Anything interesting in it's output when you try it?

```python
from falconpy import RealTimeResponse, RealTimeResponseAdmin, OAuth2, Hosts

auth = OAuth2(client_id="FALCON_CLIENT_ID",
              client_secret="FALCON_CLIENT_SECRET"
              )
hosts = Hosts(auth_object=auth)
rtr = RealTimeResponse(auth_object=auth)
rtr_admin = RealTimeResponseAdmin(auth_object=auth)

target_host="TARGET_HOSTNAME_HERE"  # You can append a wildcard here if necessary

aid = hosts.query_devices_by_filter(filter="hostname:'{target_host}'")["body"]["resources"][0]

session_id = rtr.init_session(device_id=aid)["body"]["resources"][0]["session_id"]
print(f"Session ID: {session_id}")

filename = "testfile.jpg"
testfile = [('file', (filename, open(filename, 'rb').read(), 'image/jpg'))]

upload_result = rtr_admin.create_put_files(name=filename,
                                           description="Testing file",
                                           files=testfile
                                           )
if upload_result["status_code"] == 409:
    print("File already exists in Cloud repository")
elif upload_result["status_code"] == 200:
    print("File successfully uploaded")
else:
    print("An error occurred.")


put_file_id = rtr_admin.list_put_files(filter=f"name:'{filename}'")["body"]["resources"][0]
print(f"Put file ID: {put_file_id}")

# Put the file on the target machine
cloud_request_id = rtr_admin.execute_admin_command(base_command="put",
                                                   command_string=f"put {filename}",
                                                   session_id=session_id,
                                                   persist=True
                                                   )["body"]["resources"][0]["cloud_request_id"]

print(cloud_request_id)

status = False

while not status:
    status = rtr_admin.check_admin_command_status(cloud_request_id=cloud_request_id,
                                                  sequence_id=0
                                                  )["body"]["resources"][0]["complete"]
    if status:
        print("Upload complete")
    else:
        print("Uploading")

# Cleanup our session
rtr.delete_session(session_id=session_id)
```

**Reply by usedtire** [reply]
No idea why but it's working with that code now.  I did send an email to my admin asking him to doublecheck permissions.  I really appreciate your help.  

First run with a test file:
File already exists in Cloud repository
Put file ID: f940d4f948af11ec961002f9e2b85ac3_14fdf14ccae4437dabee5e18dab2411d
88ffae84-8a9a-41f1-ab37-e57d41e2d9bb
Uploading
Uploading
Upload complete

Second Run with the exe I am trying to use after commenting out the upload:
Put file ID: e8287fb6659911ebb5c7029e904d98c3_405d7657b22e40aab9b18f2102890aaf
a0901460-ac9b-459f-8af6-b0be74bdcd48
Uploading
Uploading
Uploading
Upload complete

Thanks again
